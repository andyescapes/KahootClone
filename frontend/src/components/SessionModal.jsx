import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Modal from '@material-ui/core/Modal';
import Backdrop from '@material-ui/core/Backdrop';
import Button from '@material-ui/core/Button';
import Fade from '@material-ui/core/Fade';
import PropTypes from 'prop-types';

const useStyles = makeStyles((theme) => ({
  modal: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  paper: {
    backgroundColor: theme.palette.background.paper,
    border: '2px solid #000',
    boxShadow: theme.shadows[5],
    padding: theme.spacing(2, 4, 3),
  },
}));

export default function SessionModal (props) {
  const classes = useStyles();
  const [open, setOpen] = React.useState(true);

  const handleClose = (e) => {
    e.stopPropagation();
    setOpen(false);
    props.setter(false);
  };

  return (
    <div>
      {open && (
        <>
          <Modal
            aria-labelledby="transition-modal-title"
            aria-describedby="transition-modal-description"
            className={classes.modal}
            open={open}
            onClose={handleClose}
            closeAfterTransition
            BackdropComponent={Backdrop}
            BackdropProps={{
              timeout: 500,
            }}
          >
            <Fade in={open}>
              <div className={classes.paper}>
                <h2 id="transition-modal-title">{props.message}</h2>
                <p id="transition-modal-description">{props.data}</p>
                <Button
                  color="primary"
                  variant="contained"
                  onClick={props.onClickFn}
                  data-test-target="Results"
                >
                  {props.buttonMessage}
                </Button>
              </div>
            </Fade>
          </Modal>
        </>
      )}
    </div>
  );
}

SessionModal.propTypes = {
  message: PropTypes.string,
  setter: PropTypes.func,
  data: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  field: PropTypes.string,
  buttonMessage: PropTypes.string,
  onClickFn: PropTypes.func,
}
