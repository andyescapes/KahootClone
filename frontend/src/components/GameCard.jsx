import React from 'react';
import '../App.css';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import CardMedia from '@material-ui/core/CardMedia';
import Box from '@material-ui/core/Box';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import ErrorPopUp from '../components/ErrorPopUp';
import SessionModal from '../components/SessionModal';
import { useHistory } from 'react-router-dom';
import { makeStyles } from '@material-ui/core';
import PropTypes from 'prop-types';

function GameCard (props) {
  const history = useHistory();

  const [error, setError] = React.useState(false);
  const [sessionId, setSessionId] = React.useState('');
  const [activeSession, setActiveSession] = React.useState('');
  const [gameStopped, setGameStopped] = React.useState(false);
  const [message, setMessage] = React.useState(false);

  // get the details of a particular quiz
  const getQuizDetails = async (token, id) => {
    const request = await fetch(`http://localhost:5546/admin/quiz/${id}`, {
      headers: {
        Authorization: `Bearer ${token}`,
        'Content-Type': 'application/json',
      },
    });
    const result = await request.json();
    setSessionId(result.active);
    setActiveSession(true);
  };

  // handle calls fo start, stop and advance
  const gameApiCall = async (token, id, command) => {
    const request = await fetch(
      `http://localhost:5546/admin/quiz/${id}/${command}`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${token}`,
        },
      }
    );
    const result = await request.json();
    if (result.error) setError(result.error);

    if (request.status === 200) {
      if (command === 'start') {
        getQuizDetails(token, id);
      } else if (command === 'advance') {
        setMessage('The Quiz moved on to the next Question');
      } else {
        setGameStopped(true);
      }
    }
  };

  // error checking to see if a last game exists or not
  const endGameOnClick = () => {
    if (sessionId) {
      history.push(`/results/${props.id}/${sessionId}`);
    } else {
      setMessage('You have not finished a game since you signed in');
    }
  };

  const startGameOnClick = () => {
    navigator.clipboard.writeText(`http://localhost:3000/play/${sessionId}`);
  };

  // slight styled component
  const useStyles = makeStyles((theme) => ({
    root: {
      fontSize: '0.7em',
    },
  }));
  const classes = useStyles();
  return (
    <div>
      <Box mt ={1} mb={1}>
      <Card>
        <CardActionArea
          onClick={() =>
            props.dashboard
              ? history.push(`/edit/${props.id}`)
              : history.push(`/edit/${props.id}/${props.questionId}`)
          }
        >
          {props.thumbnail ?? (
            <CardMedia

              component="img"
              height="140"
              image={props.thumbnail}
              title="Contemplative Reptile"
            />
          )}

          <CardContent>
            <Typography gutterBottom variant="h5" component="h2">
              {props.title}
              <svg
                onClick={(event) => {
                  event.stopPropagation();
                  if (props.dashboard) {
                    props.delete(
                      props.token,
                      props.id,
                      props.setQuizzes,
                      props.quizzes
                    );
                  } else {
                    props.delete(props.token, props.id, props.questionId);
                  }
                }}
                className="exit"
                height="329pt"
                viewBox="0 0 329.26933 329"
                width="329pt"
                xmlns="http://www.w3.org/2000/svg"
                aria-label="Close Card"
              >
                <path d="m194.800781 164.769531 128.210938-128.214843c8.34375-8.339844 8.34375-21.824219 0-30.164063-8.339844-8.339844-21.824219-8.339844-30.164063 0l-128.214844 128.214844-128.210937-128.214844c-8.34375-8.339844-21.824219-8.339844-30.164063 0-8.34375 8.339844-8.34375 21.824219 0 30.164063l128.210938 128.214843-128.210938 128.214844c-8.34375 8.339844-8.34375 21.824219 0 30.164063 4.15625 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921875-2.089844 15.082031-6.25l128.210937-128.214844 128.214844 128.214844c4.160156 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921874-2.089844 15.082031-6.25 8.34375-8.339844 8.34375-21.824219 0-30.164063zm0 0" />
              </svg>
            </Typography>

            {props.details ?? (
              <Typography variant="body2" color="textSecondary" component="p">
                {props.details}
              </Typography>
            )}
          </CardContent>
        </CardActionArea>{' '}
        {props.dashboard && (
          <>
            <CardActions>
              <Button
                className={classes.root}
                size="small"
                color="primary"
                variant="contained"
                onClick={(event) => {
                  event.stopPropagation();
                  gameApiCall(props.token, props.id, 'start');
                }}
                data-test-target="Start"
              >
                Start Game
              </Button>
              <Button
                className={classes.root}
                size="small"
                color="secondary"
                variant="contained"
                onClick={(event) => {
                  event.stopPropagation();
                  gameApiCall(props.token, props.id, 'end');
                }}
                data-test-target="Stop"
              >
                Stop Game
              </Button>{' '}
              <Button
                className={classes.root}
                size="small"
                color="primary"
                variant="contained"
                onClick={(event) => {
                  event.stopPropagation();
                  gameApiCall(props.token, props.id, 'advance');
                }}
              >
                Advance Question
              </Button>
              <Button
                className={classes.root}
                size="small"
                color="primary"
                variant="contained"
                onClick={(event) => {
                  event.stopPropagation();
                  if (sessionId) {
                    history.push(`/results/${props.id}/${sessionId}`);
                  } else {
                    setMessage(
                      'You have not finished a game since you signed in'
                    );
                  }
                }}
              >
                Last Results
              </Button>
            </CardActions>
          </>
        )}
      </Card>
      {error && <ErrorPopUp setError={setError} error={error}></ErrorPopUp>}
      {activeSession && (
        <SessionModal
          setter={setActiveSession}
          data={sessionId}
          message={'New Session Started'}
          buttonMessage={'Copy Link'}
          onClickFn={startGameOnClick}
        ></SessionModal>
      )}
      {gameStopped && (
        <SessionModal
          setter={setGameStopped}
          message={'Would you like to view the results?'}
          buttonMessage={'View Results'}
          onClickFn={endGameOnClick}
        ></SessionModal>
      )}
      {message && (
        <ErrorPopUp title="" setError={setMessage} error={message}></ErrorPopUp>
      )}
      </Box>
    </div>
  );
}

GameCard.propTypes = {
  id: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  dashboard: PropTypes.bool,
  questionId: PropTypes.string,
  thumbnail: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool
  ]),
  token: PropTypes.string,
  setQuizzes: PropTypes.func,
  details: PropTypes.string,
  title: PropTypes.string,
  delete: PropTypes.func,
  quizzes: PropTypes.array
}

export default GameCard;
