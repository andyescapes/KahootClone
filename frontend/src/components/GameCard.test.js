import { shallow } from "enzyme";
import React from "react";
import renderer from "react-test-renderer";
import GameCard from "./GameCard";
import SessionModal from "./SessionModal";
import ErrorPopUp from "./ErrorPopUp";
import Typography from "@material-ui/core/Typography";
import Card from "@material-ui/core/Card";
import Button from "@material-ui/core/Button";

describe("GameCard", () => {
  const noop = () => {};

  it("contains a svg", () => {
    const wrapper = shallow(<GameCard />);
    expect(wrapper.find("svg").length).toBe(1);
  });

  it("contains a Material UI Typography", () => {
    const wrapper = shallow(<GameCard />);
    expect(wrapper.find(Typography).length).toBe(2);
  });

  it("contains a Material UI Card", () => {
    const wrapper = shallow(<GameCard />);
    expect(wrapper.find(Card).length).toBe(1);
  });

  it("contains no buttons if no dashboard prop supplied", () => {
    const wrapper = shallow(<GameCard />);
    expect(wrapper.find(Button).length).toBe(0);
  });

  it("contains a Material UI Button when dashboard prop supplied", () => {
    const wrapper = shallow(<GameCard dashboard={true} />);
    expect(wrapper.find(Button).length).toBe(4);
  });

  it("contains start game Button when dashboard prop supplied", () => {
    const wrapper = shallow(<GameCard dashboard={true} />);
    expect(wrapper.find(Button).at(0).text()).toBe("Start Game");
  });

  it("contains a stop game UI Button when dashboard prop supplied", () => {
    const wrapper = shallow(<GameCard dashboard={true} />);
    expect(wrapper.find(Button).at(1).text()).toBe("Stop Game");
  });

  it("contains an advance question Button when dashboard prop supplied", () => {
    const wrapper = shallow(<GameCard dashboard={true} />);
    expect(wrapper.find(Button).at(2).text()).toBe("Advance Question");
  });

  it("contains a last results Button when dashboard prop supplied", () => {
    const wrapper = shallow(<GameCard dashboard={true} />);
    expect(wrapper.find(Button).at(3).text()).toBe("Last Results");
  });

  it("triggers onClick event handler which calls a function", () => {
    const deleteFn = jest.fn();
    const event = {
      stopPropagation: () => {
        console.log("test");
      },
    };

    const wrapper = shallow(<GameCard dashboard={true} delete={deleteFn} />);
    wrapper.find("svg").simulate("click", event);
    expect(deleteFn).toHaveBeenCalledTimes(1);
  });

  it("contains the title supplied via props", () => {
    const wrapper = shallow(<GameCard title={"hello"} />);
    expect(wrapper.find(Typography).at(0).text()).toBe("hello");
  });

  it("by default contains no SessionModal", () => {
    const wrapper = shallow(<GameCard title={"hello"} />);
    expect(wrapper.find(SessionModal).length).toBe(0);
  });

  it("by default contains no ErorPopUp", () => {
    const wrapper = shallow(<GameCard title={"hello"} />);
    expect(wrapper.find(ErrorPopUp).length).toBe(0);
  });

  it("renders with minimal props", () => {
    const gameCard = renderer.create(<GameCard setState={noop} />).toJSON();
    expect(gameCard).toMatchSnapshot();
  });

  it("renders with provided details (snapshot)", () => {
    const gameCard = renderer
      .create(<GameCard delete={noop} details={"a detail"} />)
      .toJSON();
    expect(gameCard).toMatchSnapshot();
  });

  it("renders with additional buttons when dashboard prop active (snapshot)", () => {
    const gameCard = renderer
      .create(<GameCard delete={noop} details={"a detail"} dashboard={true} />)
      .toJSON();
    expect(gameCard).toMatchSnapshot();
  });
});
