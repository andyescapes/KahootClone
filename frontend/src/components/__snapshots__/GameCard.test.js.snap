// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameCard renders with additional buttons when dashboard prop active (snapshot) 1`] = `
<div>
  <div
    className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <button
      className="MuiButtonBase-root MuiCardActionArea-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <img
        className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
        height="140"
        title="Contemplative Reptile"
      />
      <div
        className="MuiCardContent-root"
      >
        <h2
          className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
        >
          <svg
            className="exit"
            height="329pt"
            onClick={[Function]}
            viewBox="0 0 329.26933 329"
            width="329pt"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m194.800781 164.769531 128.210938-128.214843c8.34375-8.339844 8.34375-21.824219 0-30.164063-8.339844-8.339844-21.824219-8.339844-30.164063 0l-128.214844 128.214844-128.210937-128.214844c-8.34375-8.339844-21.824219-8.339844-30.164063 0-8.34375 8.339844-8.34375 21.824219 0 30.164063l128.210938 128.214843-128.210938 128.214844c-8.34375 8.339844-8.34375 21.824219 0 30.164063 4.15625 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921875-2.089844 15.082031-6.25l128.210937-128.214844 128.214844 128.214844c4.160156 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921874-2.089844 15.082031-6.25 8.34375-8.339844 8.34375-21.824219 0-30.164063zm0 0"
            />
          </svg>
        </h2>
        a detail
      </div>
      <span
        className="MuiCardActionArea-focusHighlight"
      />
    </button>
     
    <div
      className="MuiCardActions-root MuiCardActions-spacing"
    >
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Start Game
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-containedSizeSmall MuiButton-sizeSmall"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Stop Game
        </span>
      </button>
       
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Advance Question
        </span>
      </button>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Last Results
        </span>
      </button>
    </div>
  </div>
  
</div>
`;

exports[`GameCard renders with minimal props 1`] = `
<div>
  <div
    className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <button
      className="MuiButtonBase-root MuiCardActionArea-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <img
        className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
        height="140"
        title="Contemplative Reptile"
      />
      <div
        className="MuiCardContent-root"
      >
        <h2
          className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
        >
          <svg
            className="exit"
            height="329pt"
            onClick={[Function]}
            viewBox="0 0 329.26933 329"
            width="329pt"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m194.800781 164.769531 128.210938-128.214843c8.34375-8.339844 8.34375-21.824219 0-30.164063-8.339844-8.339844-21.824219-8.339844-30.164063 0l-128.214844 128.214844-128.210937-128.214844c-8.34375-8.339844-21.824219-8.339844-30.164063 0-8.34375 8.339844-8.34375 21.824219 0 30.164063l128.210938 128.214843-128.210938 128.214844c-8.34375 8.339844-8.34375 21.824219 0 30.164063 4.15625 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921875-2.089844 15.082031-6.25l128.210937-128.214844 128.214844 128.214844c4.160156 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921874-2.089844 15.082031-6.25 8.34375-8.339844 8.34375-21.824219 0-30.164063zm0 0"
            />
          </svg>
        </h2>
        <p
          className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
        />
      </div>
      <span
        className="MuiCardActionArea-focusHighlight"
      />
    </button>
     
  </div>
  
</div>
`;

exports[`GameCard renders with provided details (snapshot) 1`] = `
<div>
  <div
    className="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <button
      className="MuiButtonBase-root MuiCardActionArea-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <img
        className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img"
        height="140"
        title="Contemplative Reptile"
      />
      <div
        className="MuiCardContent-root"
      >
        <h2
          className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
        >
          <svg
            className="exit"
            height="329pt"
            onClick={[Function]}
            viewBox="0 0 329.26933 329"
            width="329pt"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="m194.800781 164.769531 128.210938-128.214843c8.34375-8.339844 8.34375-21.824219 0-30.164063-8.339844-8.339844-21.824219-8.339844-30.164063 0l-128.214844 128.214844-128.210937-128.214844c-8.34375-8.339844-21.824219-8.339844-30.164063 0-8.34375 8.339844-8.34375 21.824219 0 30.164063l128.210938 128.214843-128.210938 128.214844c-8.34375 8.339844-8.34375 21.824219 0 30.164063 4.15625 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921875-2.089844 15.082031-6.25l128.210937-128.214844 128.214844 128.214844c4.160156 4.160156 9.621094 6.25 15.082032 6.25 5.460937 0 10.921874-2.089844 15.082031-6.25 8.34375-8.339844 8.34375-21.824219 0-30.164063zm0 0"
            />
          </svg>
        </h2>
        a detail
      </div>
      <span
        className="MuiCardActionArea-focusHighlight"
      />
      <span
        className="MuiTouchRipple-root"
      />
    </button>
     
  </div>
  
</div>
`;
